src/plan.rs

```diff
@@ -0,0 +1,467 @@
+    /// The absolute timelock this plan uses
+    pub absolute_timelock: Option<LockTime>,
+    /// The relative timelock this plan uses
+    pub relative_timelock: Option<Sequence>,
+
+    pub(crate) desc_type: DescriptorType,
+}
+
+impl Plan {
+    /// Returns the witness version
+    pub fn witness_version(&self) -> Option<WitnessVersion> {
+        self.desc_type.segwit_version()
+    }
+
+    /// The weight, in witness units, needed for satisfying this plan (includes both
+    /// the script sig weight and the witness weight)
+    pub fn satisfaction_weight(&self) -> usize {
+        self.witness_size() + self.scriptsig_size() * 4
+    }
+
+    /// The size in bytes of the script sig that satisfies this plan
+    pub fn scriptsig_size(&self) -> usize {
+        match (self.desc_type.segwit_version(), self.desc_type) {
+            // Entire witness goes in the script_sig
+            (None, _) => witness_size(self.template.as_ref()),
+            // Taproot doesn't have a "wrapped" version (scriptSig len (1))
+            (Some(WitnessVersion::V1), _) => 1,
+            // scriptSig len (1) + OP_0 (1) + OP_PUSHBYTES_20 (1) + <pk hash> (20)
+            (_, DescriptorType::ShWpkh) => 1 + 1 + 1 + 20,
+            // scriptSig len (1) + OP_0 (1) + OP_PUSHBYTES_32 (1) + <script hash> (32)
+            (_, DescriptorType::ShWsh) | (_, DescriptorType::ShWshSortedMulti) => 1 + 1 + 1 + 32,
+            // Native Segwit v0 (scriptSig len (1))
+            __ => 1,
+        }
+    }
+
+    /// The size in bytes of the witness that satisfies this plan
+    pub fn witness_size(&self) -> usize {
+        if let Some(_) = self.desc_type.segwit_version() {
+            witness_size(self.template.as_ref())
+        } else {
+            0 // should be 1 if there's at least one segwit input in the tx, but that's out of
+              // scope as we can't possibly know that just by looking at the descriptor
+        }
+    }
+}
+
+/// The Assets we can use to satisfy a particular spending path
+#[derive(Debug, Default)]
+pub struct Assets {
+    keys: HashMap<hash160::Hash, DefiniteDescriptorKey>,
+    tap_key_spend_sigs: HashMap<hash160::Hash, bitcoin::SchnorrSig>,
+    ecdsa_signatures: HashMap<hash160::Hash, (bitcoin::EcdsaSig, usize)>,
+    schnorr_signatures: HashMap<(hash160::Hash, TapLeafHash), (bitcoin::SchnorrSig, usize)>,
+    sha256_preimages: HashSet<sha256::Hash>,
+    hash256_preimages: HashSet<hash256::Hash>,
+    ripemd160_preimages: HashSet<ripemd160::Hash>,
+    hash160_preimages: HashSet<hash160::Hash>,
+    absolute_timelock: Option<LockTime>,
+    relative_timelock: Option<Sequence>,
+}
+
+impl AssetProvider<DefiniteDescriptorKey> for Assets {
+    fn lookup_ecdsa_sig(&self, pk: &DefiniteDescriptorKey) -> bool {
+        let pkh = pk.to_pubkeyhash(SigType::Ecdsa);
+        let we_have_pk = self.keys.contains_key(&pkh);
+        let we_have_sig = self.ecdsa_signatures.contains_key(&pkh);
+        we_have_pk || we_have_sig
```

Comment:
Should all of these `||`s be `&&`s? I don't understand the intended logic here -- it seems like "do we have a pk" and "do we have a sig" are two independent questions.

Reply:
Yeah it's an `OR` because in both cases we can satisfy that node: either we have the key (so we can produce the signature ourselves) or we have a pre-made signature that we found for example in a psbt.

Reply:
Ah, gotcha. Maybe worth adding a comment on one of these to explain the context.